File: app/src/main/java/spryrocks/com/tristimer/presentation/ui/screens/NavigationActivity.java
package spryrocks.com.tristimer.presentation.ui.screens;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import com.spryrocks.android.modules.utils.Actions;

import spryrocks.com.tristimer.R;
import spryrocks.com.tristimer.databinding.NavigationActivityBinding;
import spryrocks.com.tristimer.presentation.ui.screens.results.ResultsFragment;
import spryrocks.com.tristimer.presentation.ui.screens.statistics.StatisticsFragment;
import spryrocks.com.tristimer.presentation.ui.screens.timer.TimerFragment;

public class NavigationActivity extends AppCompatActivity {
    public NavigationViewModel viewModel = new NavigationViewModel();

    FragmentTransaction transaction;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = ViewModelProviders.of(this).get(NavigationViewModel.class);
        NavigationActivityBinding binding = DataBindingUtil.setContentView(this, R.layout.navigation_activity);
        binding.setView(this);
        binding.setModel(viewModel.model);

        viewModel.model.menuItem.addCallback(onMenuItemChanged);

        addFragment(new TimerFragment(), false);
    }

    public final BottomNavigationView.OnNavigationItemSelectedListener onNavigationItemSelectedListener = new BottomNavigationView.OnNavigationItemSelectedListener() {
        @Override
        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
            switch (item.getItemId()) {
                case R.id.results_item:
                    viewModel.model.menuItem.set(NavigationModel.MenuItem.RESULTS);
                    break;
                case R.id.statistics_item:
                    viewModel.model.menuItem.set(NavigationModel.MenuItem.STATISTICS);
                    break;
                case R.id.timer_item:
                    viewModel.model.menuItem.set(NavigationModel.MenuItem.TIMER);
                    break;
                default:
                    viewModel.model.menuItem.set(null);
            }
            return true;
        }
    };

    private final Actions.Action1<NavigationModel.MenuItem> onMenuItemChanged = new Actions.Action1<NavigationModel.MenuItem>() {
        @Override
        public void run(NavigationModel.MenuItem menuItem) {
            if (menuItem == null)
                return;
            else {
                switch (menuItem) {
                    case RESULTS:
                        addFragment(new ResultsFragment(), true);
                        break;
                    case STATISTICS:
                        addFragment(new StatisticsFragment(), true);
                        break;
                    case TIMER:
                        addFragment(new TimerFragment(), true);
                        break;
                }
            }
        }
    };

    public void addFragment(Fragment fragment, boolean addToBackStack) {
        transaction = getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.content, fragment, null);
        if (addToBackStack)
            transaction.addToBackStack(null);
        transaction.commit();
    }
}

File: app/src/main/java/spryrocks/com/tristimer/presentation/ui/screens/NavigationActivity.kt
package spryrocks.com.tristimer.presentation.ui.screens

import android.databinding.DataBindingUtil
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v4.app.Fragment
import spryrocks.com.tristimer.R
import spryrocks.com.tristimer.databinding.NavigationActivityBinding
import android.arch.lifecycle.ViewModelProviders
import com.spryrocks.android.modules.utils.Actions.Action1
import spryrocks.com.tristimer.presentation.ui.screens.NavigationModel.MenuItem.*
import spryrocks.com.tristimer.presentation.ui.screens.results.ResultsFragment
import spryrocks.com.tristimer.presentation.ui.screens.statistics.StatisticsFragment
import spryrocks.com.tristimer.presentation.ui.screens.timer.TimerFragment

class NavigationActivity : AppCompatActivity() {
    private var viewModel: NavigationViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val viewModel = ViewModelProviders.of(this).get(NavigationViewModel::class.java)
        this.viewModel = viewModel

        val binding: NavigationActivityBinding = DataBindingUtil.setContentView(this, R.layout.navigation_activity)
        binding.view = this
        binding.model = viewModel.model

        viewModel.model.menuItem.addCallback(onMenuItemChanged)

        addFragment(TimerFragment(), false)
    }

    val onNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item ->
        viewModel!!.model.menuItem.set(when (item.itemId) {
            R.id.results_item -> RESULTS
            R.id.statistics_item -> STATISTICS
            R.id.timer_item -> TIMER
            else -> null
        })
        true
    }

    private val onMenuItemChanged = Action1<NavigationModel.MenuItem> { menuItem ->
        if (menuItem == null)
            return@Action1
        addFragment(when (menuItem) {
            RESULTS -> ResultsFragment()
            STATISTICS -> StatisticsFragment()
            TIMER -> TimerFragment()
        })
    }

    private fun addFragment(fragment: Fragment, addToBackStack: Boolean = true) {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.content, fragment, null)
        if (addToBackStack)
            transaction.addToBackStack(null)
        transaction.commit()
    }
}
